macros/src/main/scala-3/enumeratum/ValueEnumMacros.scala

```diff
@@ -0,0 +1,265 @@
+  */
+@SuppressWarnings(Array("org.wartremover.warts.StringPlusAny"))
+object ValueEnumMacros {
+
+  /** Finds ValueEntryType-typed objects in scope that have literal `value: Int` implementations.
+    *
+    * $valueEntryTypeNote
+    */
+  def findIntValueEntriesImpl[ValueEntryType: Type](using
+      Quotes
+  ): Expr[IndexedSeq[ValueEntryType]] =
+    findValueEntriesImpl[ValueEntryType, Int]
+
+  /** Finds ValueEntryType-typed objects in scope that have literal `value: Long` implementations.
+    *
+    * $valueEntryTypeNote
+    */
+  def findLongValueEntriesImpl[ValueEntryType: Type](using
+      Quotes
+  ): Expr[IndexedSeq[ValueEntryType]] =
+    findValueEntriesImpl[ValueEntryType, Long]
+
+  /** Finds ValueEntryType-typed objects in scope that have literal `value: Short` implementations.
+    *
+    * Note
+    *
+    *   - requires the ValueEntryType to have a 'value' member that has a literal value
+    *   - the Short value should be a literal Int (do no need to cast .toShort).
+    */
+  def findShortValueEntriesImpl[ValueEntryType: Type](using
+      Quotes
+  ): Expr[IndexedSeq[ValueEntryType]] =
+    findValueEntriesImpl[ValueEntryType, Short]
+
+  /** Finds ValueEntryType-typed objects in scope that have literal `value: String` implementations.
+    *
+    * Note
+    *
+    *   - requires the ValueEntryType to have a 'value' member that has a literal value
+    */
+  def findStringValueEntriesImpl[ValueEntryType: Type](using
+      Quotes
+  ): Expr[IndexedSeq[ValueEntryType]] =
+    findValueEntriesImpl[ValueEntryType, String]
+
+  /** Finds ValueEntryType-typed objects in scope that have literal `value: Byte` implementations.
+    *
+    * Note
+    *
+    *   - requires the ValueEntryType to have a 'value' member that has a literal value
+    */
+  def findByteValueEntriesImpl[ValueEntryType: Type](using
+      Quotes
+  ): Expr[IndexedSeq[ValueEntryType]] =
+    findValueEntriesImpl[ValueEntryType, Byte]
+
+  /** Finds ValueEntryType-typed objects in scope that have literal `value: Char` implementations.
+    *
+    * Note
+    *
+    *   - requires the ValueEntryType to have a 'value' member that has a literal value
+    */
+  def findCharValueEntriesImpl[ValueEntryType: Type](using
+      Quotes
+  ): Expr[IndexedSeq[ValueEntryType]] =
+    findValueEntriesImpl[ValueEntryType, Char]
+
+  private given ValueOfFromExpr[T <: Singleton](using Type[T]): FromExpr[ValueOf[T]] with {
+    def unapply(x: Expr[ValueOf[T]])(using q: Quotes): Option[ValueOf[T]] = {
+      import q.reflect.*
+
+      x match {
+        case '{ new ValueOf[T]($v) } =>
+          v.asTerm match {
+            case id: Ident => {
+              val cls         = Class.forName(id.symbol.fullName + '$')
+              val moduleField = cls.getFields.find(_.getName == f"MODULE$$")
+
+              moduleField.map { field =>
+                new ValueOf(field.get(null).asInstanceOf[T])
```

Comment:
`field.get(null).asInstanceOf[T]` is interesting; @cchantep can you please shed some light on this?

Ultimately this is could just be my ignorance of the Scala3 macro world, but it's odd to see two "Scala taboos" consecutively.